presto-common/src/main/java/com/facebook/presto/common/type/TypeSignatureUtils.java

```diff
@@ -0,0 +1,158 @@
+    {
+        Map<TypeSignature, TypeSignature> typeSignatureMap = getTypeSignatureMap(actualTypeSignatureParameters, expectedTypeSignatureParameters);
+        return resolveTypeSignatures(actualTypeSignature, typeSignatureMap).getTypeSignature();
+    }
+
+    public static Map<TypeSignature, TypeSignature> getTypeSignatureMap(List<TypeSignature> parameters, List<TypeSignature> argumentTypes)
+    {
+        Map<TypeSignature, TypeSignature> typeSignatureMap = new HashMap<>();
+        if (argumentTypes.size() != parameters.size()) {
+            throw new IllegalStateException(
+                    "Parameters size: " + parameters.size() + " and argumentTypes size: " + argumentTypes.size() + " do not match !");
+        }
+
+        for (int i = 0; i < argumentTypes.size(); i++) {
+            TypeSignature parameter = parameters.get(i);
+            TypeSignature argumentType = argumentTypes.get(i);
+
+            // Realistically, there are only two cases,
+            // 1. When parameter.getParameters() is empty :
+            //    - Eg: parameter type = generic type(T) and argumentType = array(double), we can directly put
+            //          map.put(parameter, argumentType) there is no need to loop over the argumentType.
+            //    - Eg: param type = non-generic type(bigint) , in this case argumentType will also be of the
+            //          same non-generic type hence  map.put(parameter, argumentType) is valid here too.
+            // 2. When parameter.getParameters() is not empty:
+            //    - Eg: parameter type = generic type (array(T)) and argumentType = array(double), we recursively run
+            //      this function until we reach condition 1.
+            //      Example calls for parameter type = generic type (array(T)) and argumentType = array(double)
+            //         Iteration 1:
+            //            Parameter:
+            //              Type signature base = array
+            //              parameters = T
+            //            ArgumentType :
+            //              Type signature base = array
+            //              parameters = double
+            //         Iteration 2:
+            //            Parameter:
+            //              Type signature base = T
+            //              parameters = empty
+            //            ArgumentType :
+            //              Type signature base = double
+            //              parameters = empty
+            //          return typeSignatureMap = {"T": "double"}
+
+            // If parameter params are of type long e.g decimal(15, 2) or
+            // of type varchar e.g decimal(i4, i5), we don't need to recursively call the function on its params
+            if (parameter.getParameters().isEmpty() || areParametersLongOrVariable(parameter.getParameters())) {
```

Comment:
what about parameters that are none of these kinds (long_enum, varchar_enum, distinct_type)?

also, since you've been looking at these different parameter kinds, would be great to add some javadoc to the enum to explain what the different parameter kinds mean.